### PR TITLE
I saw there was no option to plot bar graphs

### PR DESCRIPTION
**Pull request by Bliss3d**
So I wrote quickly wrote it. Im not sure Im in compliance with vispy project standarts, probably not. So let me know if there are some changes needed.
Best Regards

### CODE REVIEW

**Comment by djhoese** [comment]
Coooool! Do you have a screenshot of what the result looks like? Should this be added to the plotting interface too (PlotWidget)?

**Comment by djhoese** [comment]
Flake8 is failing with:

```
 vispy/visuals/__init__.py:49:1: F401 '.bar.BarVisual' imported but unused
vispy/visuals/__init__.py:49:27: W292 no newline at end of file
vispy/visuals/bar.py:7:1: F401 'time' imported but unused
vispy/visuals/bar.py:9:1: F401 'datetime.date' imported but unused
vispy/plot/tests/test_plot.py:13:1: E302 expected 2 blank lines, found 1
vispy/plot/tests/test_plot.py:46:1: E302 expected 2 blank lines, found 1
vispy/plot/tests/test_plot.py:79:5: F841 local variable 'bar1' is assigned to but never used
vispy/plot/tests/test_plot.py:90:1: E303 too many blank lines (3)
```

**Comment by Bliss3d** [comment]
vispy_bar_graph

**Comment by Bliss3d** [comment]
> Coooool! Do you have a screenshot of what the result looks like? Should this be added to the plotting interface too (PlotWidget)?

I dont know Im not that familiar with the vispy. I just need a fast trading candle ploter so I needed bar graphs - like this

vispy4


However I have not looked into the PlotWidget

**Comment by djhoese** [comment]
You can see in the lint test that it is still mad at you:

```
 Running flake8... 
vispy/visuals/__init__.py:49:1: F401 '.bar.BarVisual' imported but unused
vispy/plot/tests/test_plot.py:10:1: F401 'vispy.app' imported but unused
Failed: flake8 failed
```

Interesting screenshots. I guess I assumed something more bar chart-y like:

image

I suppose that is also possible. As for the plot widget, it looks like we already have a HistogramVisual:

https://github.com/vispy/vispy/blob/182f87d62baa7158db96d963c7da435b723178cc/vispy/plot/plotwidget.py#L165-L188

I suppose it would be pretty similar to do a `.bar` like matplotlib: https://matplotlib.org/stable/plot_types/basic/bar.html#sphx-glr-plot-types-basic-bar-py

File: vispy/visuals/bar.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        Bottom of the bars
+    height : int | float | array-like
+        Height of the bars
+    width : int | float
+        Width of all bars
+    color : instance of Color
+        Color of the histogram.
+    """
+
+    def __init__(self, color='w', index=None, bottom=None, height=None, width=0.8):
```
Comment (djhoese) [inline vispy/visuals/bar.py:28]: It looks like these defaults are never handled for `index`, `bottom`, or `height`. Should these be positional arguments? Or have different defaults?

**Comment by almarklein** [comment]
>  I guess I assumed something more bar chart-y like:

Perhaps this should be called "candle" instead.

**Comment by Bliss3d** [comment]
Well the bottom is configurable, so it would probebly be smart to set the bottom default to 0 so it would be more like a traditional bar chart

File: vispy/visuals/bar.py (hunk 1 above)
Comment (Bliss3d) [reply vispy/visuals/bar.py:28]: Well index and height would have to be supplied to graph the bars, and as suggested bottom should probely 0 as default

**Comment by Bliss3d** [comment]
More like this
vispy_bar_traditional

**Comment by Bliss3d** [comment]
> > I guess I assumed something more bar chart-y like:
> 
> Perhaps this should be called "candle" instead.

Well its not relly a candle its 6 different bars chart overlayed


`bar1 = scene.Bar(index=df['Index'][:10].to_numpy(), bottom=df['Bar1Bottom'][:10].to_numpy(),
                           height=df['Bar1Height'][:10], width=0.8, color=(0.25, 0.8, 0.),
                           parent=vb1.scene)
    bar2 = scene.Bar(index=df['Index'][:10].to_numpy(), bottom=df['Bar2Bottom'][:10].to_numpy(),
                           height=df['Bar2Height'][:10], width=0.3, color=(0.25, 0.8, 0.),
                           parent=vb1.scene)
    bar3 = scene.Bar(index=df['Index'][:10].to_numpy(), bottom=df['Bar3Bottom'][:10].to_numpy(),
                           height=df['Bar3Height'][:10], width=0.3, color=(0.25, 0.8, 0.),
                           parent=vb1.scene)
    bar4 = scene.Bar(index=df['Index'][:10].to_numpy(), bottom=df['Bar4Bottom'][:10].to_numpy(),
                           height=df['Bar4Height'][:10], width=0.8, color=(1, 0.1, 0.),
                           parent=vb1.scene)
    bar5 = scene.Bar(index=df['Index'][:10].to_numpy(), bottom=df['Bar5Bottom'][:10].to_numpy(),
                           height=df['Bar5Height'][:10], width=0.3, color=(1, 0.1, 0.),
                           parent=vb1.scene)
    bar6 = scene.Bar(index=df['Index'][:10].to_numpy(), bottom=df['Bar6Bottom'][:10].to_numpy(),
                           height=df['Bar6Height'][:10], width=0.3, color=(1, 0.1, 0.),
                           parent=vb1.scene)`

**Comment by Bliss3d** [comment]
I was just confused, now everything should be fine, also code quality is now improved

**Comment by brisvag** [comment]
Trying to pick this up! In general, this looks awesome @Bliss3d :)

One thing I noticed is that the spacing between bars seems to be inconsistent at certain zoom levels. I suppose there is some rounding error? Not sure if this is an issue with the mesh shader or the code here.

image

File: vispy/visuals/bar.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    rr[1::6, X] += (1 - width) / 2
+    rr[2::6, X] += 1 - (1 - width) / 2
+    rr[3::6, X] += (1 - width) / 2
+    rr[4::6, X] += 1 - (1 - width) / 2
+    rr[5::6, X] += 1 - (1 - width) / 2
+    rr[:, Y] = np.repeat(bottom, 6)
+    rr[::2, Y] += np.repeat(height, 3)
+    index.astype(np.float32)
+
+    return rr
```
Comment (brisvag) [inline vispy/visuals/bar.py:152]: I noticed that calling `Bar.meshdata.get_vertices()` errors out. It turns out that `mesh` only half-works if you give it a `(n, 3)` array of vertices without faces (as it actually expects a `(n, 3, 3)` array of explicit faces by coordinates). For some reason it *mostly* goes silently, but some things clearly don't work.

The easiest way to solve it, I think, is to reshape the return of this function:

```
return rr.reshape(-1, 3, 3)
```

Alternatively, go the [`Histogram`](https://github.com/vispy/vispy/blob/main/vispy/visuals/histogram.py) approach and explicitly pass an array to indicate the faces.

**Review by brisvag — CHANGES_REQUESTED** [review]
Small changes to fix an issue with `MeshData`, but otherwise I think this looks good.

I agree that it would be great to add this to the plot api, but that can also be done in a followup PR, as far as I'm concerned!

**Comment by brisvag** [comment]
> yes there seems to be an issue, however i didnt mind it. i think i included a parameter for the bar spacing maybe try setting it to 0 helps to find the origin

As far as I can tell, this is just due to pixel rounding in the mesh shader. Not sure if this can/should be solved here... I think it's ouside the scope of this PR.

**Comment by Bliss3d** [comment]
it think so... However I'm not sure what exactly happend. If its ok for everybody I'll create a new PR. I did fix the meshdata vertices/faces. Also I added candle plots.

**Comment by djhoese** [comment]
Creating a new PR is fine but please keep independent features in separate PRs. It makes them easier to review and merge.

**Comment by Bliss3d** [comment]
replaced by #2394  closing

**Pull request closed without merging**